dhcpv6/option_4rd.go

```diff
@@ -0,0 +1,178 @@
+package dhcpv6
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/u-root/u-root/pkg/uio"
+)
+
+// Opt4RD represents a 4RD option. It is only a container for 4RD_*_RULE options
+type Opt4RD Options
+
+// Code returns the Option Code for this option
+func (op *Opt4RD) Code() OptionCode {
+	return Option4RD
+}
+
+// ToBytes serializes this option
+func (op *Opt4RD) ToBytes() []byte {
+	return (*Options)(op).ToBytes()
+}
+
+// String returns a human-readable representation of the option
+func (op *Opt4RD) String() string {
+	return fmt.Sprintf("Opt4RD{%v}", (*Options)(op))
+}
+
+// ParseOpt4RD builds an Opt4RD structure from a sequence of bytes.
+// The input data does not include option code and length bytes
+func ParseOpt4RD(data []byte) (*Opt4RD, error) {
+	var opt Options
+	err := opt.FromBytes(data)
+	return (*Opt4RD)(&opt), err
+}
+
+// Opt4RDMapRule represents a 4RD Mapping Rule option
+// The option is described in https://tools.ietf.org/html/rfc7600#section-4.9
+// The 4RD mapping rules are described in https://tools.ietf.org/html/rfc7600#section-4.2
+type Opt4RDMapRule struct {
+	// Prefix4 is the IPv4 prefix mapped by this rule
+	Prefix4 net.IPNet
+	// Prefix6 is the IPv6 prefix mapped by this rule
+	Prefix6 net.IPNet
+	// EABitsLength is the number of bits of an address used in constructing the mapped address
+	EABitsLength uint8
+	// WKPAuthorized determines if well-known ports are assigned to addresses in an A+P mapping
+	// It can only be set if the length of Prefix4 + EABits > 32
+	WKPAuthorized bool
+}
+
+const (
+	// opt4RDWKPAuthorizedMask is the mask for the WKPAuthorized flag in its
+	// byte in Opt4RDMapRule
+	opt4RDWKPAuthorizedMask = 1 << 7
+	// opt4RDHubAndSpokeMask is the mask for the HubAndSpoke flag in its
+	// byte in Opt4RDNonMapRule
+	opt4RDHubAndSpokeMask = 1 << 7
+	// opt4RDTrafficClassMask is the mask for the TrafficClass flag in its
+	// byte in Opt4RDNonMapRule
+	opt4RDTrafficClassMask = 1 << 0
+)
+
+// Code returns the option code representing this option
+func (op *Opt4RDMapRule) Code() OptionCode { return Option4RDMapRule }
+
+// ToBytes serializes this option
+func (op *Opt4RDMapRule) ToBytes() []byte {
+	buf := uio.NewBigEndianBuffer(nil)
+	p4Len, _ := op.Prefix4.Mask.Size()
+	p6Len, _ := op.Prefix6.Mask.Size()
+	buf.Write8(uint8(p4Len))
+	buf.Write8(uint8(p6Len))
+	buf.Write8(op.EABitsLength)
+	if op.WKPAuthorized {
+		buf.Write8(opt4RDWKPAuthorizedMask)
+	} else {
+		buf.Write8(0)
+	}
+	if op.Prefix4.IP.To4() == nil {
+		// The API prevents us from returning an error here
```

Comment:
I hate this, but I can't imagine another solution without changing the whole API and returning `([]byte, error)` for ToBytes.

Reply:
I see this a little differently.

I think nil is as good as net.IP{0, 0, 0, 0}. I don't see you checking for that?

And then there's probably a lot of other invalid values for this, like net.IP{255, 255, 255, 255}.

Are we able to enumerate the exact list of invalid things? Or does it depend on your specific network?

I don't think this library should concern itself with validation of the values. My take is - the user is responsible. Especially because in a bunch of other places in this library, nil = net.IP{0, 0, ...}, too.

Reply:
I wasn't worried about invalid packets, I definitely agree (at least this API of) the library should allow you to craft packets that don't respect the RFC requirements. Also agree that at least treating nil = net.IP{0...} is fine and common

My worry was about unrepresentable packets. Say you put an (non-ipv4-mapped) ipv6 address in the Prefix4 field, or a 5-byte slice. Those aren't valid net.IP packets and you can't represent them in the wire format, so I think they still should raise some kind of serialization error, instead of silently serializing to zeros.